### PR TITLE
Fix segment fault for alltoall

### PR DESCRIPTION
**Pull request by pengwa**
**Description**: fix segment fault for alltoall. 

In MOE model tests, there are random issues reporting segment fault. 

The problem is AllToAll autograd function is consuming a process group object in its forward function. when the segmentfault happens, the process group param accepted in the forward function is a 'str' like "ONNX:Slice114" or some wired object pointer. 

The root cause is: during PythonOp export, we save the process group (as a non tensor pointer type) object's address only, and save it in PythonOp's attribute. But python did not know there is a reference to that object. In the new env MOE test is trying to onboard, it looks there is more possibility process group is destroyed before the tests completed. If it is teared down before PythonOp/PythonOpGrad completed, an unexpected object (e.g. the str or werid object) is picked up and used as process group type, segement fault occurs. 

The fix: the process group as inputs of all PythonOp, they should exist along the training lifetime. So we can track those non-tensor object reference once we are exporting PythonOp. Keeping those object in a global store will avoid that object be released. We will clean up the store before python program exits. For multiple UT cases, cleaning up is not triggered, but it should be fine. 

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_exporter.py [hunk 1]
```diff
@@ -111,6 +119,8 @@ def _export_pt_1_10(g, n, *args, **kwargs):
                     # All other inputs are accessed via "pointers".
                     input_pointer_scalar_positions.append(i)
                     input_pointer_scalars.append(id(arg))
+
+                    NONTENSOR_OBJECT_POINTER_STORE[id(arg)] = arg
```
Comment (wschin) [inline orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_exporter.py:123]: This can be a workaround. For a sustainable fix, do you feel we can increase its reference count here and decrease reference count in PythonOp's dtor? Or we should provie an API to release  objects in NONTENSOR_OBJECT_POINTER_STORE? For example,
```python
# training loop
for ....
  ort.start_cache_object_pointers()
  loss = model(x)
  loss.backward()
 optimizer.step()
 ...
  ort.release_pointers_cached_in_this_iteration()
```
You can use `with` as well, e.g.,
```python
for ....
  with object_pointer_pool():
    loss = model(x)
    loss.backward()
    optimizer.step()
 ...
```

Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_exporter.py:123]: Imagine pointers and other non-tensor constant arguments are "constant", so ideally once model is exported, the pointers should also be valid unless the model ends its lifetime. This is how I understand their relation. 

One usage pattern looks like this to end the scope before compute runs:

idx=0
while idx < 12:
  group=create_process_group()
  m = prepare_torch_model(group)
  m = ORTModule(m)
  loss = torch.sum(m(input))
  loss.backward()

when idx=0 loop complete it should be fine.
when idx=1, group is overridden, if the idx=0's cuda compute (PythonOp) did not complete, it will use a release buffer, which is unexpected. 

Add more scope in user script did not helps for this case.

Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_exporter.py:123]: Let PythonOp kernels maintain the ref make some senses, there is one trick issue there. Imagine PythonOp kernel creation, we increase the ref cnt by 1, in its destructor, we decrease the ref cnt by 1.  If Python program exit, but C++ did not complete destructors for those kernels, An invalid pythonobject will be released by kernel's destructors, a segement fault will happen then, unless we have some ways to trigger those releasing proactively before python program exit (using atexit).

Comment (baijumeswani) [reply orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_exporter.py:123]: Given that this arg is a non tensor type, can we make a deepcopy of it and save it to the pythonop attribute? Then the ownership of the copy will be with the pythonop.

Comment (wschin) [reply orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_exporter.py:123]: > Imagine pointers and other non-tensor constant arguments are "constant", so ideally once model is exported, the pointers should also be valid unless the model ends its lifetime. This is how I understand their relation.
> 
> One usage pattern looks like this to end the scope before compute runs:
> 
> idx=0 while idx < 12: group=create_process_group() m = prepare_torch_model(group) m = ORTModule(m) loss = torch.sum(m(input)) loss.backward()
> 
> when idx=0 loop complete it should be fine. when idx=1, group is overridden, if the idx=0's cuda compute (PythonOp) did not complete, it will use a release buffer, which is unexpected.
> 
> Add more scope in user script did not helps for this case.

I think PyTorch has the same problem with this example, so we can ignore it. PyTorch users should write PyTorch-compatible code.

Comment (wschin) [reply orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_exporter.py:123]: > Let PythonOp kernels maintain the ref make some senses, there is one trick issue there. Imagine PythonOp kernel creation, we increase the ref cnt by 1, in its destructor, we decrease the ref cnt by 1. If Python program exit, but C++ did not complete destructors for those kernels, An invalid pythonobject will be released by kernel's destructors, a segement fault will happen then, unless we have some ways to trigger those releasing proactively before python program exit (using atexit).

If use `atexit` solves the problem, it's fine, right?

**Review by wschin — APPROVED** [review]

File: orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_exporter.py (hunk 1 above)
Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_exporter.py:123]: > Given that this arg is a non tensor type, can we make a deepcopy of it and save it to the pythonop attribute? Then the ownership of the copy will be with the pythonop.

Oh, cloning a torch process group might be tricky, I am not sure whether the object is self-contained, or clone-able. Maybe we can revisit if we need such kind of fix later.

Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_exporter.py:123]: Talked with @wschin, the conclusion is, we do this way without introducing more complex in C++ backend side. A python interface is also can be used for potential usage if customers find those object references are not released in time (which should not be a common usage case).

**Comment by pengwa** [comment]
Thanks @wschin @baijumeswani.
